Added support to evict pods that use local storage.

adyada opened the pull request:
Added a command line flag to enable eviction of pods that use local storage. 

We are trying to use descheduler in our k8s cluster, but most of our pods use local storage. The local storage on a pod is mainly because of our logging pipeline, but we don't if some logs are lost while eviction. I have heard that there is a similar option to override local storage in k8s auto scaler. 

Let me know if this is a right approach to do this. Also please let me know where to add tests as I am new to go language.

aveshagarwal commented:
/ok-to-test

aveshagarwal commented:
@adyada run ./hack/update-* scripts on your code changes, and update and resubmit this PR?

towolf commented:
@adyada you gonna finish this?

adyada commented:
Okay, will try to address comments today. I got busy in some other stuff, so didn't get back to desehduler stuff.

adyada commented:
On running 
./hack/update-generated-conversions.sh

I am getting this error 

can't load package: package github.com/kubernetes-incubator/descheduler/vendor/k8s.io/code-generator/cmd/conversion-gen: cannot find package "github.com/kubernetes-incubator/descheduler/vendor/k8s.io/code-generator/cmd/conversion-gen" in any of:
	/usr/lib/go-1.6/src/github.com/kubernetes-incubator/descheduler/vendor/k8s.io/code-generator/cmd/conversion-gen (from $GOROOT)
	($GOPATH not set)


Can anyone please help regarding how to resolve this? i am new to using go language.

aveshagarwal commented:
> ($GOPATH not set)

set or export GOPATH and rerun the command.

adyada commented:
Thanks. That resolved that go error. But now I am getting another error. I will take a look, but just posting here in case you know what's wrong.

**adyada@adyada-Virtual-Machine:~/deschedulerCode/deschedulerFork/descheduler$ ./hack/update-generated-conversions.sh** 

ERROR: logging before flag.Parse: F1119 16:32:11.469102    6500 conversion.go:313] failed to find pkg: k8s.io/apimachinery/pkg/conversion
goroutine 1 [running]:
github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog.stacks(0xaa2200, 0x0, 0x0, 0x0)
	/home/adyada/gopath/src/github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog/glog.go:769 +0xb8
github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog.(*loggingT).output(0xa87480, 0xc800000003, 0xc82017a300, 0xa6c51d, 0xd, 0x139, 0x0)
	/home/adyada/gopath/src/github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog/glog.go:720 +0x2ce
github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog.(*loggingT).printf(0xa87480, 0xc800000003, 0x868090, 0x16, 0xc821739aa0, 0x1, 0x1)
	/home/adyada/gopath/src/github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog/glog.go:655 +0x1d4
github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog.Fatalf(0x868090, 0x16, 0xc821739aa0, 0x1, 0x1)
	/home/adyada/gopath/src/github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog/glog.go:1148 +0x5d
github.com/kubernetes-incubator/descheduler/vendor/k8s.io/code-generator/cmd/conversion-gen/generators.Packages(0xc820eee0c0, 0xc8202321b0, 0x0, 0x0, 0x0)
	/home/adyada/gopath/src/github.com/kubernetes-incubator/descheduler/vendor/k8s.io/code-generator/cmd/conversion-gen/generators/conversion.go:313 +0x19ea
github.com/kubernetes-incubator/descheduler/vendor/k8s.io/gengo/args.(*GeneratorArgs).Execute(0xc8202321b0, 0xc820078db0, 0x817798, 0x6, 0x901150, 0x0, 0x0)
	/home/adyada/gopath/src/github.com/kubernetes-incubator/descheduler/vendor/k8s.io/gengo/args/args.go:176 +0x321
main.main()
	/home/adyada/gopath/src/github.com/kubernetes-incubator/descheduler/vendor/k8s.io/code-generator/cmd/conversion-gen/main.go:68 +0x367


**adyada@adyada-Virtual-Machine:~/deschedulerCode/deschedulerFork/descheduler$ ./hack/update-generated-deep-copies.sh** 
ERROR: logging before flag.Parse: F1119 16:34:08.311414    6642 main.go:74] Error: Failed executing generator: some packages had errors:
type "k8s.io/apimachinery/pkg/runtime.Object" in k8s:deepcopy-gen:interfaces tag of type k8s.io/apimachinery/pkg/runtime.Object is not an interface, but: ""
type "k8s.io/apimachinery/pkg/runtime.Object" in k8s:deepcopy-gen:interfaces tag of type k8s.io/apimachinery/pkg/runtime.Object is not an interface, but: ""
type "k8s.io/apimachinery/pkg/runtime.Object" in k8s:deepcopy-gen:interfaces tag of type k8s.io/apimachinery/pkg/runtime.Object is not an interface, but: ""
type "k8s.io/apimachinery/pkg/runtime.Object" in k8s:deepcopy-gen:interfaces tag of type k8s.io/apimachinery/pkg/runtime.Object is not an interface, but: ""
goroutine 1 [running]:
github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog.stacks(0xaa2200, 0x0, 0x0, 0x0)
	/home/adyada/gopath/src/github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog/glog.go:769 +0xb8
github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog.(*loggingT).output(0xa87420, 0xc800000003, 0xc821b6ccc0, 0xa6cce5, 0x7, 0x4a, 0x0)
	/home/adyada/gopath/src/github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog/glog.go:720 +0x2ce
github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog.(*loggingT).printf(0xa87420, 0xc800000003, 0x81acc0, 0x9, 0xc820e87e88, 0x1, 0x1)
	/home/adyada/gopath/src/github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog/glog.go:655 +0x1d4
github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog.Fatalf(0x81acc0, 0x9, 0xc820e87e88, 0x1, 0x1)
	/home/adyada/gopath/src/github.com/kubernetes-incubator/descheduler/vendor/github.com/golang/glog/glog.go:1148 +0x5d
main.main()
	/home/adyada/gopath/src/github.com/kubernetes-incubator/descheduler/vendor/k8s.io/code-generator/cmd/deepcopy-gen/main.go:74 +0x327


On running "make" I get this error

**adyada@adyada-Virtual-Machine:~/deschedulerCode/deschedulerFork/descheduler$ make**
CGO_ENABLED=0 go build -ldflags "-X github.com/kubernetes-incubator/descheduler/cmd/descheduler/app.version=`git describe --tags` -X github.com/kubernetes-incubator/descheduler/cmd/descheduler/app.buildDate=`date +%FT%T%z` -X github.com/kubernetes-incubator/descheduler/cmd/descheduler/app.gitCommit=`git rev-parse HEAD`" -o _output/bin/descheduler github.com/kubernetes-incubator/descheduler/cmd/descheduler
../../../gopath/src/github.com/kubernetes-incubator/descheduler/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:20:2: cannot find package "context" in any of:
	/home/adyada/gopath/src/github.com/kubernetes-incubator/descheduler/vendor/context (vendor tree)
	/usr/lib/go-1.6/src/context (from $GOROOT)
	/home/adyada/gopath/src/context (from $GOPATH)
Makefile:33: recipe for target 'build' failed
make: *** [build] Error 1


I think, only "make image" worked for me before.  In case you know what's wrong with my setup, then please let me know. Thanks.

aveshagarwal commented:
```
 github.com/kubernetes-incubator/descheduler/pkg/descheduler/strategies
pkg/descheduler/strategies/duplicates_test.go:80:36: not enough arguments in call to deleteDuplicatePods
	have (*"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/client-go/kubernetes/fake".Clientset, string, []*"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".Node, bool, nodePodEvictedCount, number)
	want (kubernetes.Interface, string, []*"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".Node, bool, nodePodEvictedCount, int, bool)
pkg/descheduler/strategies/lownodeutilization_test.go:115:40: not enough arguments in call to classifyNodes
	have (NodePodsMap, "github.com/kubernetes-incubator/descheduler/pkg/api".ResourceThresholds, "github.com/kubernetes-incubator/descheduler/pkg/api".ResourceThresholds)
	want (NodePodsMap, "github.com/kubernetes-incubator/descheduler/pkg/api".ResourceThresholds, "github.com/kubernetes-incubator/descheduler/pkg/api".ResourceThresholds, bool)
pkg/descheduler/strategies/lownodeutilization_test.go:220:40: not enough arguments in call to classifyNodes
	have (NodePodsMap, "github.com/kubernetes-incubator/descheduler/pkg/api".ResourceThresholds, "github.com/kubernetes-incubator/descheduler/pkg/api".ResourceThresholds)
	want (NodePodsMap, "github.com/kubernetes-incubator/descheduler/pkg/api".ResourceThresholds, "github.com/kubernetes-incubator/descheduler/pkg/api".ResourceThresholds, bool)
pkg/descheduler/strategies/node_affinity_test.go:179:64: not enough arguments in call to removePodsViolatingNodeAffinityCount
	have (*options.DeschedulerServer, "github.com/kubernetes-incubator/descheduler/pkg/api".DeschedulerStrategy, string, []*"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".Node, nodePodEvictedCount, int)
	want (*options.DeschedulerServer, "github.com/kubernetes-incubator/descheduler/pkg/api".DeschedulerStrategy, string, []*"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".Node, nodePodEvictedCount, int, bool)
pkg/descheduler/strategies/pod_antiaffinity_test.go:58:44: not enough arguments in call to removePodsWithAffinityRules
	have (*"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/client-go/kubernetes/fake".Clientset, string, []*"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".Node, bool, nodePodEvictedCount, number)
	want (kubernetes.Interface, string, []*"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".Node, bool, nodePodEvictedCount, int, bool)
pkg/descheduler/strategies/pod_antiaffinity_test.go:64:43: not enough arguments in call to removePodsWithAffinityRules
	have (*"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/client-go/kubernetes/fake".Clientset, string, []*"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".Node, bool, nodePodEvictedCount, number)
	want (kubernetes.Interface, string, []*"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".Node, bool, nodePodEvictedCount, int, bool)
FAIL	github.com/kubernetes-incubator/descheduler/pkg/descheduler/strategies [build failed]
?   	github.com/kubernetes-incubator/descheduler/pkg/utils	[no test files]
?   	github.com/kubernetes-incubator/descheduler/test	[no test files]
make: *** [test-unit] Error 2
The command "make test-unit" exited with 2.
```

aveshagarwal commented:
```
$ hack/verify-gofmt.sh
!!! 'gofmt -s' needs to be run on the following files: 
./cmd/descheduler/app/options/options.go
./pkg/apis/componentconfig/v1alpha1/types.go
The command "hack/verify-gofmt.sh" exited with 1.
```

aveshagarwal commented:
@adyada see the above issues reported by travis CI.

adyada commented:
Thanks Avesh. I addressed your comments. I also ran "make test-unit" and it passed locally. ci/openshift-jenkins  is still failing for this PR. Where can I get the logs for ci/openshift-jenkins?

aveshagarwal commented:
/retest

aveshagarwal commented:
> Thanks Avesh. I addressed your comments. I also ran "make test-unit" and it passed locally. ci/openshift-jenkins is still failing for this PR. Where can I get the logs for ci/openshift-jenkins?

The logs should be in the failed test link as part of artifacts.

adyada commented:
I think,  ci/openshift-jenkins build is failing because of the following error.

Processing triggers for ureadahead (0.100.0-19) ...
Updating project ssh metadata...
....................failed.
this version of kubeadm only supports deploying clusters with the control plane version >= 1.12.0. Current version: v1.11.1
++ export status=FAILURE
++ status=FAILURE
+ set +o xtrace
########## FINISHED STAGE: FAILURE: PROVISION GCE CLUSTER [00h 07m 20s] ##########
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Executing post build scripts.

aveshagarwal commented:
Can one of the admins verify this patch?

aveshagarwal commented:
/retest

Nowaker commented:
@aveshagarwal @ravisantoshgudimetla @jayunit100 Can one of you please take a look at this and approve if all is fine, or comment on what needs to be done to get it approved and merged?

aveshagarwal commented:
Please squash commits into 2 commits: one with manual code changes, and another one with auto-generated code changes.

Nowaker commented:
Thanks. If @adyada doesn't do it by 4/17, I will.

rhockenbury commented:
@adyada Were you planning on taking care of the squash?

adyada commented:
I didn't know how to squash into two commits, so I somehow I ended up deleting earlier PR, and created a new PR with two commits https://github.com/kubernetes-incubator/descheduler/pull/142
